src/interceptor/context.rs

```diff
@@ -1,37 +1,37 @@
-use super::{Interceptor, InterceptorFuture, InterceptorObj};
-use crate::{body::AsyncBody, error::Error};
-use http::{Request, Response};
-use std::{fmt, sync::Arc};
+use super::{Interceptor, InterceptorFuture};
+use crate::{body::AsyncBody, error::Error, HttpClient};
+use http::Request;
+use std::fmt;
 
 /// Execution context for an interceptor.
-pub struct Context<'a> {
-    pub(crate) invoker: Arc<dyn Invoke + Send + Sync + 'a>,
-    pub(crate) interceptors: &'a [InterceptorObj],
+pub struct Context {
+    pub(crate) client: HttpClient,
+    pub(crate) interceptor_offset: usize,
 }
 
-impl<'a> Context<'a> {
+impl Context {
     /// Send a request asynchronously, executing the next interceptor in the
     /// chain, if any.
-    pub async fn send(&self, request: Request<AsyncBody>) -> Result<Response<AsyncBody>, Error> {
-        if let Some(interceptor) = self.interceptors.first() {
+    pub fn send(&self, request: Request<AsyncBody>) -> InterceptorFuture<Error> {
+        if let Some(interceptor) = self.client.interceptors().get(self.interceptor_offset) {
             let inner_context = Self {
-                invoker: self.invoker.clone(),
-                interceptors: &self.interceptors[1..],
+                client: self.client.clone(),
+                interceptor_offset: self.interceptor_offset + 1,
             };
 
-            interceptor.intercept(request, inner_context).await
+            interceptor.intercept(request, inner_context)
         } else {
-            self.invoker.invoke(request).await
+            self.client.invoke(request)
         }
     }
 }
 
-impl fmt::Debug for Context<'_> {
+impl fmt::Debug for Context {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("Context").finish()
     }
 }
 
 pub(crate) trait Invoke {
-    fn invoke(&self, request: Request<AsyncBody>) -> InterceptorFuture<'_, Error>;
+    fn invoke(&self, request: Request<AsyncBody>) -> InterceptorFuture<Error>;
```

Comment:
Since we simply reference `HttpClient` directly inside of `Context`, we can probably remove this trait and make the implementation on `HttpClient` an inherent method instead. There was never any plan to make it possible to implement custom `Invoke`s, it was merely used as a means of creating a trait object.

Reply:
Addressed.